#                                 Shortcomings of Avg Order Value (AOV)

<B>  IMPORTING THE REQUIRED LIBRARIES </B>

In [55]:
import pandas as pd
import numpy as np
from datetime import datetime
import matplotlib.pyplot as plt
from scipy.stats import mode

<B> READING DATA FROM THE <a href="https://docs.google.com/spreadsheets/d/16i38oonuX1y1g7C_UAmiK9GkY7cS-64DfiDMNiR41LM/edit#gid=0" target="_blank">GOOGLE SHEET</a> PROVIDED </B>

In [2]:
url = "https://docs.google.com/spreadsheets/d/16i38oonuX1y1g7C_UAmiK9GkY7cS-64DfiDMNiR41LM/edit#gid=0"

url_to_read = url.replace("/edit#gid=", "/export?format=csv&gid=")

data = pd.read_csv(url_to_read)

<B> TAKING A LOOK AT THE DATA </B>

In [4]:
data.head(5)

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0         1       53      746           224            2           cash   
1         2       92      925            90            1           cash   
2         3       44      861           144            1           cash   
3         4       18      935           156            1    credit_card   
4         5       18      883           156            1    credit_card   

            created_at  
0  2017-03-13 12:36:56  
1  2017-03-03 17:38:52  
2   2017-03-14 4:23:56  
3  2017-03-26 12:43:37  
4   2017-03-01 4:35:11

- Checking for Null Values in the given data

In [6]:
data.isna().sum().sum()

0

- Checking the total number of rows present in the given data

In [7]:
len(data)

5000

- Creating a new column that stores dates in the Datetime format instead of strings

In [8]:
data['created_date'] = ''

for i in range(len(data)):
    
    data['created_date'][i] = datetime.strptime(data['created_at'][i][0:10], '%Y-%m-%d')

<ipython-input-8-155832fa811e>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['created_date'][i] = datetime.strptime(data['created_at'][i][0:10], '%Y-%m-%d')


- Sorting the data in a chronological order

In [9]:
df = data.sort_values(by = 'created_date')

<B> CALCULATING THE AVERAGE ORDER VALUE </B>

In [40]:
np.average(df['order_amount'])

3145.128

<B> TAKING A LOOK AT THE SUMMARY STATISTICS FOR THE 'order_amount' VARIABLE IN OUR DATA </B>

In [10]:
df['order_amount'].describe()

count      5000.000000
mean       3145.128000
std       41282.539349
min          90.000000
25%         163.000000
50%         284.000000
75%         390.000000
max      704000.000000
Name: order_amount, dtype: float64

- We notice that the **distribution for the variable 'order_amount' is _skewed_**.
 
  
- This is evident from the fact that the **difference between the third quartile and the max value is unnaturally high**. 
 
  
- Computing statistics such as **mean or weighted mean would be misleading as the presence of outliers affects the mean disproportionately**

Since most of the observations (atleast 75%) have an order amount less than 1000, **we check for the true number of observations that have an order amount _greater_ than 1000 Dollars**

In [44]:
len(df[df['order_amount'] > 1000])

71

- <B>71/5000 Observations consist of an order amount greater than $1000</B>
 
  
- **This is less than 2 percent of the total observations**
 
  
- However, **we don't discard these outliers** as they might offer some valuable insights into the behaviour of particular customers and stores
 
  
- Instead, _we use a different measure of central tendency_ - <U>**The Mode**</U>
 
  
- The Mode, unlike the mean, is not affected by outliers. Mode is the value that has the highest frequency in the distribution
 
  
- **In our case, the mode would denote the most likely amount of revenue generated from an order**

In [59]:
mode(df['order_amount'])

ModeResult(mode=array([153], dtype=int64), count=array([87]))

- The mode for this distribution is **153**
 
  
- A basic inference would be to assume that an incoming order is more likely to generate a revenue of 153$

In [ ]:
np.median(df['order_amount'])

In [60]:
len(df[(df['order_amount'] >= 100) & (df['order_amount'] <= 200)])

1765

In [61]:
len(df[(df['order_amount'] >= 200) & (df['order_amount'] <= 300)])

855

In [63]:
len(df[(df['order_amount'] >= 300) & (df['order_amount'] <= 400)])

1166

In [62]:
len(df[(df['order_amount'] >= 0) & (df['order_amount'] <= 100)])

43

In [41]:
# df[df['order_amount'] == 704000]

In [42]:
# df[df['user_id'] == 607]

In [ ]:
# The user bearing the userid 607 has made peculiar transactions (all outliers) worth looking into

In [43]:
# df[df['order_amount'] > 1500]['shop_id'].unique()

In [ ]:
# The stores bearing the shop ids 78 and 42 are the only stores involved in unnaturally high amount values for orders, 
# contributing to almost all of the 71 outliers 